### PR TITLE
[Php81] Do not add final modifier on class constant that the class has children on FinalizePublicClassConstantRector

### LINKED ISSUE #6973: Incorrect behavior of FinalizePublicClassConstantRector
# Bug Report

| Subject        | Details              |
| :------------- | :--------------------|
| Rector version | 0.12.15      |
| Installed as   | composer dependency  |

## Minimal PHP Code Causing Issue

See https://getrector.org/demo/1ec7f537-6973-6358-8246-11db50a94ec6

```php
<?php

abstract class ParentClass
{
    public const TEST_CONSTANT = 'value1';
}

class ChildClass extends ParentClass
{
    public const TEST_CONSTANT = 'value2';
}
```

### Responsible rules

* `FinalizePublicClassConstantRector`

## Expected Behavior

I think "final" should not be added. Because we want to be able to override the constant in this case.

### PR DESCRIPTION
**Pull request by samsonasik**
Closes #1738 
Fixes https://github.com/rectorphp/rector/issues/6973

### CODE REVIEW

File: rules/Php81/Rector/ClassConst/FinalizePublicClassConstantRector.php [hunk 1]
```diff
@@ -22,13 +25,16 @@
...
+        private readonly ReflectionProvider $reflectionProvider,
+        private readonly ClassAnalyzer $classAnalyzer,
         private readonly VisibilityManipulator $visibilityManipulator
     ) {
     }
 
     public function getRuleDefinition(): RuleDefinition
     {
-        return new RuleDefinition('Add final to constants that', [
+        return new RuleDefinition('Add final to constants that does not has children', [
```
Comment (TomasVotruba) [inline rules/Php81/Rector/ClassConst/FinalizePublicClassConstantRector.php]:
```suggestion
        return new RuleDefinition('Add final to constants that does not have children', [
```

**Comment by samsonasik** [comment]
I am merging it ;)

**Comment by TomasVotruba** [comment]
Thanks :+1: 

This change still depends on the order of classes, as they're analyzed, as PHPStan and Rector process files in chunk, not as a whole.

If the child class will be discovered after, the parent will be still finalized. Yet, there is not much we can do about it now.

**Comment by samsonasik** [comment]
@TomasVotruba for overlapped file order troubleshooting, that documented in the docs https://github.com/rectorphp/rector/blob/main/docs/static_reflection_and_autoload.md#troubleshooting

**Pull request merged**